### PR TITLE
[Whisper] 🚨 Fix whisper decoding 🚨

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

This PR finalizes #30984 which enabled short-form (_<=30sec_) and long-form generation using temperature fallback. Indeed, the original OpenAI implementation uses the same [decode_with_fallback](https://github.com/openai/whisper/blob/25639fc17ddc013d56c594bfbf7644f2185fad84/whisper/transcribe.py#L279) for both short-form and long-form audio, while we used to have temperature fallback only for long-form audio.

It aims to solve issues and divergences with the original implementation: 
1. When decoding with timestamps, the Transformers implementation skips the last segment for short-form audio while the OpenAI implementation does not: it will use the sliding window strategy and go to the last generated timestamp (so indeed no differences between inferring short-form and long-form audio). This was not detected in the tests since [test_tiny_timestamp_generation](https://github.com/huggingface/transformers/blob/37ea04013b34b39c01b51aeaacd8d56f2c62a7eb/tests/models/whisper/test_modeling_whisper.py#L2048) and [test_large_timestamp_generation](https://github.com/huggingface/transformers/blob/37ea04013b34b39c01b51aeaacd8d56f2c62a7eb/tests/models/whisper/test_modeling_whisper.py#L2185) are incorrect. 
2. Miscalculation of the `avg_logprobs` that triggered temperature fallback when it should not.
3. When decoding short-form / long-form audio, we returned:
   1. **short-form**`decoder_input_ids + predicted tokens (including the eos token)` 
   2. **long-form** only the `predicted tokens (without eos token)` 
Since short-form and long-form generation are now merged, we need a consistent way of returning outputs (at least for the tokens, we still need to differentiate for past_key_values see here). To be consistent with generate convention and since the tokenizer has the skip_special_tokens argument, I went for option 1.

##  🚨 Important changes 🚨
 ➡️  **Previously:**
	•	Short-form: Returned a `ModelOutput` or `torch.LongTensor`, including decoder input IDs and the EOS token ID.
	•	Long-form: Returned a `Dict` or `torch.LongTensor`, excluding decoder input IDs and the EOS token ID.

➡️  **From now on:**
Short-form and long-form generation are now treated identically, meaning output differentiation based on these modes is no longer applicable.

Decoder input IDs and EOS token IDs are never returned, except in two specific cases: when `return_dict_in_generate=True` and (`return_timestamps=False` or `force_unique_generate_call=True`).

In this case, the output will be a `ModelOutput`, which is the result of the underlying call to GenerationMixin’s generate. Indeed, `return_timestamps=False` ensures no seeking occurs; only a single call to generate is made. Therefore, this output includes both decoder input IDs and the EOS token ID.

## Testing

> [!NOTE]  
> This PR reconciles our implementation with OpenAI's. It should therefore be tested with the updated and verified tests introduced in #34111

## Evaluations

Let’s verify the effectiveness of this fix. We will evaluate both accuracy and inference speed. To do this, we will test on four short-form test sets and two long-form test sets (effectively filtering samples as ≤30 sec and >30 sec). For the short-form sets, we will compare results with the current main branch of Transformers on the first 100 samples. (Indeed, Issue #2, which involves miscalculation of the avg_logprobs, causes the full test set to be too slow.)

⚠️ Moreover, it was also tested with [this PR](https://github.com/huggingface/transformers/pull/34537) merged. 

As explained in [this PR](https://github.com/huggingface/transformers/pull/34111), we need to use a simple Whisper fork to ensure consistency with the same input features.

> [!IMPORTANT]
> **TL;DR:** We achieve perfect 1-to-1 matching in prediction results with greedy decoding for both short-form and long-form samples, validating that our implementation matches OpenAI’s original decoding algorithm. For short-form, this PR is ~5x faster than the current implementation (due to multiple incorrect temperature fallbacks).❗

<details>
  <summary>Results 📊</summary>
  
### → short-form (first 100 samples)

Set 1: [edinburghcstr/ami](https://huggingface.co/datasets/edinburghcstr/ami), config `"ihm"`, split `test[:100]`
Set 2: [distil-whisper/chime4](https://huggingface.co/datasets/distil-whisper/chime4), config `"1-channel"`, split `test[:100]`
Set 3: [google/fleurs](https://huggingface.co/datasets/google/fleurs), config `"en_us"`, split `test[:100]`

Wandb results [here](https://wandb.ai/eustache-lebihan-hf/benchmark-whisper-short)!

|                                       | set 1 - WER | set 1 - RTFx | set 1 - WER | set 1 - RTFx | set 1 - WER | set 1 - RTFx |
| ------------------------------------- | ----------- | ------------ | ----------- | ------------ | ----------- | ------------ |
| **whisper-orig ([script](https://gist.github.com/eustlb/1828634ea898b77e47ebdaa38268cf83/2eb34add5400aa376e596b1cb8e576ea09c52cd2))**      | 17.35       | 6.55         | 4.62        | 12.21        | 3.78        | 14.61        |
| **whisper-orig-fork ([script](https://gist.github.com/eustlb/465cbe9319627089ab127b404012b9f4/44260c3aa0017f0d3776f9259322298d096c7dc0))** | 19.4        | 6.45         | 4.55        | 12           | 3.78        | 14.12        |
| **this PR ([script](https://gist.github.com/eustlb/1fa238d17d4f1f49b19b79c834fb5c7d/c31de96fd7e9d6388219c2cb621063352b2fcea6))**           | 19.4        | 3.49         | 4.55        | 8.05         | 3.78        | 10.59        |
| **currrent main ([script](https://gist.github.com/eustlb/1fa238d17d4f1f49b19b79c834fb5c7d/ed5d5d18c8e97eb8416631506a448ce4e50b8fa1))**     | 25.8        | 0.73         | 5.67        | 1.49         | 7.09        | 1.9          |</details>

### → short-form (full test sets)

Set 1: [edinburghcstr/ami](https://huggingface.co/datasets/edinburghcstr/ami), config `"ihm"`, split `test`
Set 2: [distil-whisper/chime4](https://huggingface.co/datasets/distil-whisper/chime4), config `"1-channel"`, split `test`
Set 3: [google/fleurs](https://huggingface.co/datasets/google/fleurs), config `"en_us"`, split `test`

Wandb results [here](https://wandb.ai/eustache-lebihan-hf/benchmark-whisper-short)!

|                       | **Set 1 - WER** | **Set 1 - RTFx** | **Set 2 - WER** | **Set 2 - RTFx** | **Set 3 - WER** | **Set 3 - RTFx** |
| --------------------- | --------------- | ---------------- | --------------- | ---------------- | --------------- | ---------------- |
| **whisper-orig ([script](https://gist.github.com/eustlb/1828634ea898b77e47ebdaa38268cf83/185f8c8e8e90e087c548232dad43e1010a3c5d31))**      | 16.22           | 7.86             | 10.74           | 11.45            | 4.21            | 13.79            |
| **whisper-orig-fork ([script](https://gist.github.com/eustlb/465cbe9319627089ab127b404012b9f4/b21c03011b7d8bcf9af7f937bfa35439608fb052))** | 16.36           | 7.91             | 11.08           | 11.51            | 4.14            | 13.90            |
| **this PR ([script](https://gist.github.com/eustlb/1fa238d17d4f1f49b19b79c834fb5c7d/819ec3d768df5d2e41b9367b7cc3301330476d02))**           | 16.36           | 4.18             | 11.08           | 7.51             | 4.14            | 10.67            |

### → long-form 

Set 1: [distil-whisper/tedlium-long-form](https://huggingface.co/datasets/distil-whisper/tedlium-long-form), config `"default"`, split `test`
Set 2: [distil-whisper/meanwhile](https://huggingface.co/datasets/distil-whisper/meanwhile), config `"default"`, split `test`

Wandb results [here](https://wandb.ai/eustache-lebihan-hf/benchmark-whisper-long)!

|                                       | **Set 1 - WER** | **Set 1 - RTFx** | **Set 2 - WER** | **Set 2 - RTFx** |
| ------------------------------------- | --------------- | ---------------- | --------------- | ---------------- |
| **whisper-orig ([script](https://gist.github.com/eustlb/1828634ea898b77e47ebdaa38268cf83/4619d515e03708464050aa0f0b205536ea6497db))**      | 172.15          | 4.98             | 264.66          | 4.09             |
| **whisper-orig-fork ([script](https://gist.github.com/eustlb/465cbe9319627089ab127b404012b9f4/8c717feb0c9ea98e0849945c78b62c381ce94069))** | 172.15          | 4.83             | 264.66          | 4.00             |
| **this PR ([script](https://gist.github.com/eustlb/1fa238d17d4f1f49b19b79c834fb5c7d/52cba2a9e8109f022418b9874b5791717abb4284))**           | 172.15          | 4.88             | 264.66          | 3.98             |
| **current main ([script](https://gist.github.com/eustlb/1fa238d17d4f1f49b19b79c834fb5c7d/9a343e35d28ea39d7dac8f39d5faffafbb4e613f))**     | 172.15          | 4.92             | 264.66          | 3.98             |

### CODE REVIEW

**Comment by eustlb** [comment]
**Correction:**
After discussion, it has been decided to rather go for solution 2: return only the generated tokens and skip all the special tokens.
Pros: fewer changes, no ambiguities. 
Cons: need to overwrite generic tests of GenerationMixin

**Comment by ArthurZucker** [comment]
Sounds good!

File: src/transformers/models/whisper/generation_whisper.py [hunk 1]
```diff
@@ -567,11 +567,14 @@ def generate(
         # 3. Retrieve logits processors
         device = kwargs["encoder_outputs"][0].device if "encoder_outputs" in kwargs else input_features.device
         begin_index = init_tokens.shape[1]
+        num_beams = (
+            generation_config.num_beams if generation_config.num_beams is not None else kwargs.get("num_beams", 1)
+        )
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:572]: If I remember correctly, the kwargs arguments should have priority over the generation config, which is not the case here.

cc @ArthurZucker could you confirm ?

File: src/transformers/models/whisper/generation_whisper.py [hunk 2]
```diff
@@ -905,8 +897,16 @@ def generate_with_fallback(
...
                     temperature,
+                    is_first_segment,
+                    decoder_input_ids,
                 )
 
+                # remove eos token
+                if seek_sequence[-1] == generation_config.eos_token_id:
+                    seek_sequence = seek_sequence[:-1]
+                    if return_token_timestamps:
+                        seek_outputs[i]["token_timestamps"] = seek_outputs[i]["token_timestamps"][:-1]
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:908]: Could you explain why we still remove the eos token here ?

File: tests/models/whisper/test_modeling_whisper.py [hunk 3]
```diff
@@ -1631,6 +1632,208 @@ def test_labels_sequence_max_length_error_after_changing_config(self):
...
+
+    def _contrastive_generate(self, model, inputs_dict, **kwargs):
+        if not hasattr(model, "_generate_patched"):
+            model.generate = types.MethodType(self.generate_patch(model.generate.__func__), model)
+            model._generate_patched = True
+        return super()._contrastive_generate(model, inputs_dict, **kwargs)
+
+    # Whisper's generate method does return the decoder initial tokens.
+    # We need to adapt GenerationTesterMixin's test_generate_continue_from_past_key_values to account for this.
+    # Adapted from https://github.com/huggingface/transformers/blob/67890de3b86c81fb4775f41b4690b2abaf2a19cf/tests/generation/test_utils.py#L1720
```
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:1701]: (nit) no need to say "adapted from", it's already implied.
```suggestion
```

File: tests/models/whisper/test_modeling_whisper.py [hunk 4]
```diff
@@ -1631,6 +1632,208 @@ def test_labels_sequence_max_length_error_after_changing_config(self):
...
+            # 2. generate to max length (which can be achieved by setting the eos token to an invalid value), which
+            #    would make the test flaky (e.g. EOS is generated on iteration 1 on both generations, but the
+            #    continuation would force it to generate beyond an EOS token)
+            # 3. ignore `token_type_ids` for simplicity
+            # 4. ignore `forced_eos_token_id`, which requires further manipulation of the continuation inputs and is
+            #    active by default on some models
+            # 5. ignore `encoder_no_repeat_ngram_size`, which is set by default in some encoder-decoder models. When
+            #    we use their decoder as a stand-alone model, `encoder_no_repeat_ngram_size` actually prevents
+            #    repetition exclusively from the prompt. This test relies on comparing one call vs 2 calls
+            #    with cache, what is considered a prompt is different in the two cases.
```
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:1718]: (nit) should we get rid of this to improve readability? WDYT ?

```suggestion
```

File: tests/models/whisper/test_modeling_whisper.py [hunk 5]
```diff
@@ -1631,6 +1632,208 @@ def test_labels_sequence_max_length_error_after_changing_config(self):
...
+            for layer_idx in range(len(outputs_cached.past_key_values)):
+                for kv_idx in range(len(outputs_cached.past_key_values[layer_idx])):
+                    self.assertTrue(
+                        torch.allclose(
+                            outputs.past_key_values[layer_idx][kv_idx],
+                            outputs_cached.past_key_values[layer_idx][kv_idx],
+                        )
+                    )
+
+    # Whisper's generate method does return the decoder initial tokens.
```
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:1769]: Same comment here!

**Review by ylacombe — APPROVED** [review]
Thanks for working on this @eustlb, LGTM!

The way you adapted the generation tests is quite elegant IMO, let's hear if the core maintainers agree.

Nice work as well on the benchmark results. 

If we sum them up, we're now aligned with the original Whisper implementation in terms of WER. However, short-form transcription is still considerably slower. This is not an issue in itself, but the fact that the RTFx is the same for long-form transcription seems to indicate that we could bridge the gap as well for short-form transcription. What are your thoughts on this?


Note: Supersedes #33917

**Comment by ylacombe** [comment]
Let's make sure the last failing tests are fixed before asking for @ArthurZucker's review!

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/whisper/generation_whisper.py:573]: yep! 
```suggestion
        num_beams = kwargs.get("num_beams", generation_config.get("num_beams", 1))
        )
```
rather no? (default to kwargs if wargs is not None

File: src/transformers/models/whisper/generation_whisper.py [hunk 6]
```diff
@@ -1764,10 +1800,10 @@ def _retrieve_avg_logprobs(scores, tokens, eos_token_id, temperature):
         logprobs = F.log_softmax((scores * rescale_temperature).float(), dim=-1).to(scores.dtype)
 
         # retrieve logprob of selected tokens and sum
-        sum_logprobs = sum((logprobs[i][tokens[i]] * (tokens[i] != eos_token_id)) for i in range(logprobs.shape[0]))
-        length = (tokens != eos_token_id).sum(-1) if eos_token_id is not None else tokens.shape[0]
+        # don't remove the eos token logprob! it counts in avg_logprob calculation in the original implementation
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:1803]: good catch I guess!

File: src/transformers/models/whisper/generation_whisper.py [hunk 7]
```diff
@@ -1850,3 +1886,30 @@ def _retrieve_segment(
             segment_offset = seek_num_frames[prev_idx]
 
         return segments, segment_offset
+
+    @staticmethod
+    def generate_add_decoder_input_ids(func):
+        """
+        Wrapper to add decoder input token ids to the generated sequence.
+        Whisper's generate method does not return the decoder input token ids.
+        This wrapper can be used to add the decoder input token ids to the generated sequence.
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:1895]: nice explanation!

File: src/transformers/models/whisper/generation_whisper.py [hunk 8]
```diff
@@ -616,6 +624,23 @@ def generate(
...
+
+            assistant_model = kwargs["assistant_model"]
+            if assistant_model is self:
+                assistant_model = copy.deepcopy(kwargs["assistant_model"])
+            if (
+                not hasattr(assistant_model, "_generate_with_input_tokens")
+                and type(assistant_model) != WhisperForCausalLM
+            ):
+                assistant_model.generate = types.MethodType(
+                    self.generate_add_decoder_input_ids(assistant_model.generate.__func__), assistant_model
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:639]: # we need to return decoder input ids to make sure assistant generated tokens are the same as the shape of the logits from the forward path

or something to explain why we need to do this! 🤗 
Also really not a fan of non explicit solutions like this one, tho I am guessing there are not many other things you can do!

File: tests/models/whisper/test_modeling_whisper.py [hunk 9]
```diff
@@ -193,6 +193,31 @@ def prepare_whisper_inputs_dict(
     }
 
 
+# Whisper's generate method does not return the decoder input token ids.
+# To keep compatibility with generic GenerationTesterMixin, we'll need to patch generate with this decorator.
+def with_decoder_input_ids(func):
```
Comment (ArthurZucker) [inline tests/models/whisper/test_modeling_whisper.py:198]: I don't mind it for testing purposes !

**Review by ArthurZucker — APPROVED** [review]
Thanks, would have enjoyed a solution without wrappers around generate as IMO it's too magic. Explicit is usually how we go. Tho I don't think this is a blocker, unless you did not try other solution (fine with generate modifications! 👀 ) 😉

File: src/transformers/models/whisper/generation_whisper.py (hunk 8 above)
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py:652]: added in [c9b36c3](https://github.com/huggingface/transformers/pull/34135/commits/c9b36c3a5a299f5a5f7fa9c3eadeef0f04b417d5)

File: src/transformers/models/whisper/generation_whisper.py [hunk 10]
```diff
@@ -619,6 +633,30 @@ def generate(
             generation_config=generation_config,
         )
 
+        # 5bis speculative decoding: patch generate to make sure the assistant model returns the input tokens
+        # Whisper's generate method does not return the decoder input token ids.
+        # In Transformers, assisted decoding logic does a call to the generate method of the assistant model and expects the decoder input token ids to be returned.
+        # This patch is used to make sure the assistant model returns the decoder input token ids.
+        if "assistant_model" in kwargs:
+            from .modeling_whisper import WhisperForCausalLM
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:641]: we don't really need to import this no? we can check the str or `__class__`!

File: src/transformers/models/whisper/generation_whisper.py [hunk 11]
```diff
@@ -619,6 +633,30 @@ def generate(
...
+            if assistant_model is self:
+                assistant_model = copy.deepcopy(kwargs["assistant_model"])
+            if (
+                not hasattr(
+                    assistant_model, "_generate_with_input_tokens"
+                )  # make sure the assistant model does not already have the patch
+                and type(assistant_model)
+                != WhisperForCausalLM  # such a patch is not necessary for WhisperForCausalLM that uses directly GenerationMixin's generate
+            ):
+                # patch
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/generation_whisper.py:653]: I am guessing we are patching because the assistant model can be another class than `WhisperForCausalLM` right?

**Review by ArthurZucker — APPROVED** [review]
one small nit and you can merge 🔥

File: src/transformers/models/whisper/generation_whisper.py [hunk 12]
```diff
@@ -145,31 +148,72 @@ def _pad_to_max_length(
...
+                token_timestamps_list.append(torch.tensor([], device=device))
 
     max_total_length = cut_off_length + 1 if padding == "max_length" else max_total_length
     for i in range(len(current_segments)):
         pad_length = max_total_length - len(sequences[i])
         pad = (0, pad_length) if padding_side == "right" else (pad_length, 0)
+
         sequences[i] = F.pad(sequences[i], pad=pad, value=pad_token_id)
+        if return_token_timestamps:
+            token_timestamps_list[i] = F.pad(token_timestamps_list[i], pad=pad, value=token_timestamps_list[i][-1])
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:208]: There's a case in which we append an empty list to `token_timestamps_list`, aren't we going to raise an error if `i` refers to this empty list?

File: src/transformers/models/whisper/generation_whisper.py [hunk 13]
```diff
@@ -438,21 +482,33 @@ def generate(
...
-                else if the passed input is <= 30 seconds / >= 3000 mel input features, the possible [`~utils.ModelOutput`] types are:
+                The possible [`~utils.ModelOutput`] types are:
+                - [`~utils.GenerateEncoderDecoderOutput`]
+                - [`~utils.GenerateBeamEncoderDecoderOutput`]
 
-                    - [`~generation.GenerateEncoderDecoderOutput`],
-                    - [`~generation.GenerateBeamEncoderDecoderOutput`]
+                A passed input requires only one call to the underlying `generate` method:
+                    - necessarily if `return_timestamps=False` (and necessarily the passed input is <= 30 seconds / <= 3000 mel input features since otherwise `return_timestamps` needs to be set to `True`).
+                    - eventualy if `return_timestamps=True` and the passed input is <= 30 seconds / <= 3000 mel input features, if the last predicted timestam is not < 30 seconds which will trigger a second call to generate starting to this timestamp.
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:496]: (nit)
```suggestion
                    - eventually if `return_timestamps=True` and the passed input is <= 30 seconds / <= 3000 mel input features, if the last predicted timestamp is not < 30 seconds which will trigger a second call to generate starting to this timestamp.
```

File: src/transformers/models/whisper/generation_whisper.py [hunk 14]
```diff
@@ -438,21 +482,33 @@ def generate(
...
+                - `idxs`: the start(included) and end indices (excluded) of the `tokens` of the segment in the generated sequence ids (present in `result`).
+                - `result`: the result of underlying call to `generate`.
+
+                When `return_segments=True`, `return_dict_in_generate=True` applies to each `result` of each `segment`.
 
         Example:
 
-        - *Longform transcription*: To transcribe or translate audios longer than 30 seconds, process the audio files without truncation and pass all mel features at once to generate.
+        - *Longform transcription*: To transcribe or translate audios longer than 30 seconds, process the audio files without truncation and pass all mel features at once to generate. It is necessary to set `return_timestamps=True`.
+        Indeed, long-form transcription uses a sequential algorithm based on timestamps predictions, with heuristics like compression ratio threshold, log probability threshold and temperature fallback. This algorithm is described in the [the Whisper original paper](https://cdn.openai.com/papers/whisper.pdf), section *3.8. Long-form Transcription*.
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:511]: Nice!

File: src/transformers/models/whisper/generation_whisper.py [hunk 15]
```diff
@@ -438,21 +482,33 @@ def generate(
...
-            or when `config.return_dict_in_generate=True`) or a `torch.FloatTensor` or a dict of segments when `return_segments=True`.
+            [`~utils.ModelOutput`] or `Dict[str, Any]` or `torch.LongTensor`:
 
-                If the passed input is > 30 seconds / > 3000 mel input features and `return_segments=True` then a dictionary of generated sequence ids, called `sequences` and a list of each generated segment is returned.
+                A [`~utils.ModelOutput`] (when `return_dict_in_generate=True` and the passed input requires only one call to the underlying `generate` method) or a `Dict[str, Any]` (when `return_segments=True` or `return_token_timestamps=True`, with keys `sequences`, `segments` if `return_segments=True`, `token_timestamps` if `return_token_timestamps=True`;
+                `sequences` and `token_timestamps` includes the decoder input ids and end of sequence id if the passed input requires only one call to the underlying `generate` method) or a torch.LongTensor` (including the decoder input ids and end of sequence id if the passed input requires only one call to the underlying `generate` method).
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:488]: I think this part would benefit from being clearer: don't hesitate to use bullet points and spaces, and avoid parentheses.

File: src/transformers/models/whisper/generation_whisper.py (hunk 13 above)
Comment (ylacombe) [reply src/transformers/models/whisper/generation_whisper.py:496]: I'm not sure to follow. There's two if conditions, how are they linked ? Is it ",and if the last..." or ", or if the last..." ?

File: src/transformers/models/whisper/generation_whisper.py [hunk 16]
```diff
@@ -438,21 +482,33 @@ def generate(
...
+                - [`~utils.GenerateBeamEncoderDecoderOutput`]
 
-                    - [`~generation.GenerateEncoderDecoderOutput`],
-                    - [`~generation.GenerateBeamEncoderDecoderOutput`]
+                A passed input requires only one call to the underlying `generate` method:
+                    - necessarily if `return_timestamps=False` (and necessarily the passed input is <= 30 seconds / <= 3000 mel input features since otherwise `return_timestamps` needs to be set to `True`).
+                    - eventualy if `return_timestamps=True` and the passed input is <= 30 seconds / <= 3000 mel input features, if the last predicted timestam is not < 30 seconds which will trigger a second call to generate starting to this timestamp.
 
-                else only the generated output sequence ids are returned.
+                `segments` is a list of lists (one per batch element) of `segment`.
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:498]: I assume that the length of the list is equal to the number of calls to the internal `generate` method?

File: tests/models/whisper/test_modeling_whisper.py [hunk 17]
```diff
@@ -1990,7 +1787,7 @@ def test_tiny_generation(self):
         generated_ids = model.generate(input_features, num_beams=5, max_length=20)
         transcript = processor.tokenizer.decode(generated_ids[0])
 
-        EXPECTED_TRANSCRIPT = " Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel"
+        EXPECTED_TRANSCRIPT = "<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel"
```
Comment (ylacombe) [inline tests/models/whisper/test_modeling_whisper.py:1790]: As discussed offline, I don't think we should have behavior that depends on an internal call over which the user has no influence.

Before, we had those extra tokens if the audio was short (<30s). Now, we add to this an internal condition: there's only one call to the internal `generate`.

IMO, we should have a consistent behaviour: removing these special tokens by default. And add an additional flag to allow this edge case.

**Review by ylacombe — COMMENTED** [review]
Hey @eustlb, thanks for iterating!

File: src/transformers/models/whisper/generation_whisper.py (hunk 16 above)
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py:498]: No, it is a list of segments, meaning a list whose length depends on the predicted timestamp tokens. The only way to know the number of calls to the internal `generate` method here is to look at the segment's `result`.
Indeed, we can have `[{start_1, end_1, result_1}, {start_2, end_2, result_1}, {start_3, end_3, result_2}]`: the two first segments come from the same call to internal generate (result_1) and the last to the next one (result_2).

**Comment by eustlb** [comment]
Thanks a lot @ylacombe for the review. I updated it based on your comments (see the updated PR comment) !

File: tests/models/whisper/test_modeling_whisper.py (hunk 17 above)
Comment (eustlb) [reply tests/models/whisper/test_modeling_whisper.py]: fixed in [0ea240f](https://github.com/huggingface/transformers/pull/34135/commits/0ea240fa88d2c7aeb3b8968427fb3f15a8e76e15) 👌

File: src/transformers/models/whisper/generation_whisper.py (hunk 15 above)
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py]: fixed in [0ea240f](https://github.com/huggingface/transformers/pull/34135/commits/0ea240fa88d2c7aeb3b8968427fb3f15a8e76e15) 👌

File: src/transformers/models/whisper/generation_whisper.py (hunk 12 above)
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py]: fixed in [0ea240f](https://github.com/huggingface/transformers/pull/34135/commits/0ea240fa88d2c7aeb3b8968427fb3f15a8e76e15) 👌

File: src/transformers/models/whisper/generation_whisper.py [hunk 18]
```diff
@@ -686,15 +687,25 @@ def generate(
...
+        if "assistant_model" in kwargs:
+            assistant_model = kwargs["assistant_model"]
+            assistant_model.generation_config.force_unique_generate_call = True
+
+        if hasattr(generation_config, "force_unique_generate_call"):
+            force_unique_generate_call = generation_config.force_unique_generate_call
+        elif hasattr(self.generation_config, "force_unique_generate_call"):
+            force_unique_generate_call = self.generation_config.force_unique_generate_call
+        else:
+            force_unique_generate_call = False
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:701]: What if `kwargs["force_unique_generate_call"]=True` ?

File: src/transformers/models/whisper/generation_whisper.py [hunk 19]
```diff
@@ -480,30 +484,27 @@ def generate(
...
                 specific kwargs should not be prefixed and decoder specific kwargs should be prefixed with *decoder_*.
-
         Return:
             [`~utils.ModelOutput`] or `Dict[str, Any]` or `torch.LongTensor`:
 
-                A [`~utils.ModelOutput`] (when `return_dict_in_generate=True` and the passed input requires only one call to the underlying `generate` method) or a `Dict[str, Any]` (when `return_segments=True` or `return_token_timestamps=True`, with keys `sequences`, `segments` if `return_segments=True`, `token_timestamps` if `return_token_timestamps=True`;
-                `sequences` and `token_timestamps` includes the decoder input ids and end of sequence id if the passed input requires only one call to the underlying `generate` method) or a torch.LongTensor` (including the decoder input ids and end of sequence id if the passed input requires only one call to the underlying `generate` method).
+                A:
+                - [`~utils.ModelOutput`] when `return_dict_in_generate=True` and `return_timestamps=False`, including the decoder input ids and end of sequence id.
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:491]: Is it accurate? I thought it was the case if force_unique_generate_call=True?

File: src/transformers/models/whisper/generation_whisper.py [hunk 20]
```diff
@@ -480,30 +484,27 @@ def generate(
...
+                - `Dict[str, Any]` when (`return_dict_in_generate=True` and `return_timestamps=True`) or `return_segments=True` or `return_token_timestamps=True`.
+                - `torch.LongTensor`, excluding the decoder input ids and end of sequence id.
```
Comment (ylacombe) [inline src/transformers/models/whisper/generation_whisper.py:493]: (maybe ?)
```suggestion
                - `torch.LongTensor`, excluding the decoder input ids and end of sequence id, in all other cases.
```

**Review by ylacombe — APPROVED** [review]
Thanks @eustlb for iterating! LGTM now!

File: src/transformers/models/whisper/generation_whisper.py (hunk 18 above)
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py]: I've added a `force_unique_generate_call` arg to generate 👌

**Review by ArthurZucker — APPROVED** [review]
THanks for removing the wrapper!

File: src/transformers/models/whisper/generation_whisper.py [hunk 21]
```diff
@@ -146,31 +148,76 @@ def _pad_to_max_length(
...
+                else segments[0]["result"]["sequences"]
+                for segments in current_segments
+            ],
+            dim=0,
+        )
+        if return_token_timestamps:
+            token_timestamps = torch.stack(
+                [segments[0]["result"]["token_timestamps"] for segments in current_segments], dim=0
+            )
+            return sequences, token_timestamps
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/whisper/generation_whisper.py]: Nice catch! fixed in [a1f4e43](https://github.com/huggingface/transformers/pull/34135/commits/a1f4e4315f21d6b8f903930c1530d5697a85bfe3)

**Comment by eustlb** [comment]
The four failing slow tests are expected! Merging 🤗
